cpp/include/cudf/join_new.hpp

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+namespace join {
+
+struct IdentityFilter {
+  bool __device__ operator()(table_device_view const& left, table_device_view const& right,
+                             cudf::size_type left_idx, cudf::size_type right_idx) {
+    return true;
+  }
+};
+
+class EqualColumnFilter {
+public:
+  EqualColumnFilter(cudf::size_type const* left_on,
+                    cudf::size_type const* right_on,
+                    cudf::size_type size): left_on_(left_on), right_on_(right_on), size_(size) {}
+  
+  bool __device__ operator()(table_device_view const& left, table_device_view const& right,
+                             cudf::size_type left_idx, cudf::size_type right_idx) {
+
+    bool reply = true;
+    for (cudf::size_type i = 0 ; reply && (i < size_) ; ++i) {
+      auto l = left.column(left_on_[i]);
+      auto r = right.column(right_on_[i]);
+
+      // TODO: need to work has_nulls and nulls_are_equal correctly in the below call
+      reply = cudf::experimental::type_dispatcher(l.type(),
+                                                  element_equality_comparator<true>(l, r, false),
+                                                  left_idx, right_idx);
+    }
+
+    return reply;
+  }
+
+private:
+  cudf::size_type const* left_on_;
+  cudf::size_type const* right_on_;
+  cudf::size_type        size_;
+};
+
+//
+//  Option 1: namespace by type of join
```

Comment:
There is an elegance to this approach as the contents of each namespace (inner_join, left_join, full_join) are exactly the same.